mcs/class/Mono.Posix/Mono.Unix.Native/Syscall.cs

```diff
@@ -1416,6 +1440,177 @@ public struct Linger {
+		}
+
+		public unsafe void CopyTo (byte[] destination, int startIndex)
+		{
+			fixed (uint* ptr = &s_addr)
+				Marshal.Copy ((IntPtr) ptr, destination, startIndex, 4);
+		}
+
+		public unsafe byte this[int index] {
+			get {
+				if (index < 0 || index >= 4)
+					throw new ArgumentOutOfRangeException ("index", "index < 0 || index >= 4");
+				fixed (uint* ptr = &s_addr)
+					return ((byte*) ptr)[index];
+			}
+			set {
+				if (index < 0 || index >= 4)
+					throw new ArgumentOutOfRangeException ("index", "index < 0 || index >= 4");
+				fixed (uint* ptr = &s_addr)
+					((byte*) ptr)[index] = value;
+			}
+		}
+
+		public override string ToString ()
+		{
+			return NativeConvert.ToIPAddress (this).ToString ();
+		}
+
+		public override int GetHashCode ()
+		{
+			return s_addr.GetHashCode ();
+		}
+		public override bool Equals (object obj)
+		{
+			if (!(obj is InAddr))
+				return false;
+			return Equals ((InAddr) obj);
+		}
+		public bool Equals (InAddr value)
+		{
+			return s_addr == value.s_addr;
+		}
+	}
+
+	[Map]
+	[StructLayout (LayoutKind.Sequential)]
+	public struct In6Addr : IEquatable<In6Addr> {
+		ulong addr0;
+		ulong addr1;
+
+		public unsafe In6Addr (byte[] buffer)
+		{
+			if (buffer.Length != 16)
+				throw new ArgumentException ("buffer", "buffer.Length != 16");
+			addr0 = addr1 = 0;
+			fixed (ulong* ptr = &addr0)
+				Marshal.Copy (buffer, 0, (IntPtr) ptr, 16);
+		}
+
+		public unsafe void CopyTo (byte[] destination, int startIndex)
+		{
+			fixed (ulong* ptr = &addr0)
+				Marshal.Copy ((IntPtr) ptr, destination, startIndex, 16);
+		}
+
+		public unsafe byte this[int index] {
+			get {
+				if (index < 0 || index >= 16)
+					throw new ArgumentOutOfRangeException ("index", "index < 0 || index >= 16");
+				fixed (ulong* ptr = &addr0)
+					return ((byte*) ptr)[index];
+			}
+			set {
+				if (index < 0 || index >= 16)
+					throw new ArgumentOutOfRangeException ("index", "index < 0 || index >= 16");
+				fixed (ulong* ptr = &addr0)
+					((byte*) ptr)[index] = value;
+			}
+		}
+
+		public override string ToString ()
+		{
+			return NativeConvert.ToIPAddress (this).ToString ();
+		}
+
+		public override int GetHashCode ()
+		{
+			return addr0.GetHashCode () ^ addr1.GetHashCode ();
+		}
+		public override bool Equals (object obj)
+		{
+			if (!(obj is In6Addr))
+				return false;
+			return Equals ((In6Addr) obj);
+		}
+		public bool Equals (In6Addr value)
+		{
+			return addr0 == value.addr0 && addr1 == value.addr1;
+		}
+	}
+
+	[Map ("struct cmsghdr")]
+	[CLSCompliant (false)]
+	public struct Cmsghdr {
+		public long cmsg_len;
+		public UnixSocketProtocol cmsg_level;
+		public UnixSocketControlMessage cmsg_type;
+
+		[DllImport (Syscall.MPH, SetLastError=true,
+				EntryPoint="Mono_Posix_Cmsghdr_getsize")]
+		static extern int getsize ();
+		static readonly int size = getsize ();
+		public static int Size {
+			get {
+				return size;
+			}
+		}
+
+		public static unsafe Cmsghdr Get (Msghdr msgh, long cmsg)
```

Comment:
> I'm not sure what Get() is supposed to be doing, or what POSIX API it corresponds to. Is it CMSG_FIRSTHDR()? CMSG_NXTHDR()? Something else entirely? (I'm inclined to say "something else entirely," but that doesn't help me in reviewing.)
> 
> A rename and/or comments may be helpful.

The method is for retrieving a `Cmsghdr` structure from a buffer. In C you don't need a method for that, you simply cast the pointer you have to `struct cmsghdr*` (in C you would have pointers instead of offsets into the buffer).
I can add a comment explaining that, but I'm not sure regarding the best name for the method. `ReadFromBuffer()`?

> 
> Looking at the corresponding unit tests...this is deep, deep, voodoo.

Yes, but I'm not sure how to avoid that (the C API is rather ugly already).